feat(language_server): use linter runtime

Sysix opened the pull request:
closes #7118

Added tests for:
- (language_server & VSCode) nested configuration
- (language_server & VSCode) nested configuration extends
- (language_server) without nested configuration

The implementation requires at the moment a feature flag for `oxc_linter`. I added documentation in `runtime.rs` to let others know why this is needed. Maybe someone can refactor it to avoid the flag. My first idea is a Trait for file system, but you guys know what to do <3 




![grafik](https://github.com/user-attachments/assets/9bb3aa67-c376-4f51-8474-1d0837d78338)

Sysix commented:
* **#10268** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10268?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10268?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10268).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camchenry commented:
Just commenting here to say I think this is generally the right direction: we should use the same runtime as `oxlint`, which should also enable us to use the cross-module analysis for import plugin and such. We'll probably need to implement a walker for getting all lintable files though. Then, the other hard part is determining which files need to be re-linted based on which file was saved.

Sysix commented:
> We'll probably need to implement a walker for getting all lintable files though. Then, the other hard part is determining which files need to be re-linted based on which file was saved.

I thought about that. My preferred way is to handle it inside the runtime. 
When oxlint uses a command like `npx oxlint /some/deep/folder/` and the files inside using imports like `../<filename>` we should still analyze them.
We can give the runtime with the paths a good hint to pre-work the files, but I do not think we can solve that completely without reading files when needed.
I found already some code in the runtime to diff between only semantics vs lint files. Maybe the goal to it is not so far away.

Sysix commented:
@camchenry I can see the finish line for the first time. '
Can you look please at the architecture if this is OK how I want to implement it :) 
- an oxc_linter Feature Flag
- Runtime linear and not using a Sender/receiver
- Clones

camchenry commented on crates/oxc_language_server/src/linter/tester.rs:
```diff
@@ -89,11 +92,39 @@ pub struct Tester<'t> {
+        }
+    }
+
+    pub fn new_with_linter_cross_module(linter: Linter) -> Self {
```
rather than adding lots of `new_*` functions, should we replace this with a `new_with_server_linter` method, and let the user pass in whatever options and linter they want?

Sysix replied:
To be honest, I do not like the current implementation of the Server Linter and co. The test setup reflects it^^.
Added your change and will look into refactoring.

camchenry commented on crates/oxc_linter/src/service/offset_to_position.rs:
```diff
@@ -0,0 +1,108 @@
+        Self { line, character: column }
+    }
+
+    pub fn line(&self) -> u32 {
+        self.line
+    }
```
do we need these as methods, or could we just make these fields `pub`?

Boshen reviewed (commented):
I'll test and release the lsp changes before merging this.

Sysix committed:
feat(language_server): use linter runtime (#10268)

closes #7118

Added tests for:
- (language_server & VSCode) nested configuration
- (language_server & VSCode) nested configuration extends
- (language_server) without nested configuration

The implementation requires at the moment a feature flag for `oxc_linter`. I added documentation in `runtime.rs` to let others know why this is needed. Maybe someone can refactor it to avoid the flag. My first idea is a Trait for file system, but you guys know what to do <3

![grafik](https://github.com/user-attachments/assets/9bb3aa67-c376-4f51-8474-1d0837d78338)
